File: app/channels/post_channel.rb
class PostChannel < ApplicationCable::Channel
  # before_action :set_text, only: [:speak]
  def subscribed
    
    stream_from "post_channel#{params['post']}"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def speak(data)
    print("=========")
    print(data) #{"text"=>{"text"=>"eeeeeeeee", "user_id"=>"1"}, "action"=>"speak"}が取れる
    #つまりpost_channel.jsの$(document).on ...のobjが同ファイルのspeak関数を通ってここにきた。

    #_text.html.erbで表示される内容に関係する。
    Text.create! content: data['text']['message'], user_id: data['text']['user_id'], post_id: data['text']['post_id'], name: data['text']['name']   #params['post']
         #上のデータはどう確認する？  #jsonの値の取り方に注意。data['user_id']だけじゃ取れない。
  end

  # def set_text
  #   @text = Text.find(params[:id])
  # end

  # # Only allow a list of trusted parameters through.
  # def text_params
  #   params.require(:text).permit(:content, :user_id, :post_id)
  # end

end

File: app/javascript/channels/post_channel.js
import { param } from "jquery";
import consumer from "./consumer"

//サーバ側のチャンネル（コントローラ的な）を購読する
const chatChannel = consumer.subscriptions.create("PostChannel", {
  connected() {
    // Called when the subscription is ready for use on the server
  },

  disconnected() {
    // Called when the subscription has been terminated by the server
  },

  received: function(data) {//dataはどこから来るの?
    console.log("#############################")
    console.log(data['text'])
    // ↑ _text.html.erbがhtmlの形で出力される
    return $('#texts').append(data['text']); //ここで新しい情報を追加している。
  },

  speak: function(text) {
  //perform メソッドにより、フロントのデータをサーバーに送信することができます。サーバサイドのspeakアクションにtextパラメータを渡す。
  //第1引数にサーバー側のメソッド名、第2引数に送信するデータを指定
  //console.log(text)
  // console.log("????????????????????????????")
    return this.perform('speak', {
      text: text //rails側で出すと
                       //speak({"text"=>{"text"=>"eeeeeeeee", "user_id"=>"1", post_id => "3"}})  こんな感じ。
                       //つまり下の$(document).on ...で書いているobjの内容が入る
      
    });
  }
});
//[data-behavior~=post_speaker]でイベントを取得するHTML要素を指定

$(document).on('keypress', '[data-behavior~=post_speaker]', function(event) {
  //フロントからのイベントを受け取る
  let obj = {}
  // console.log(event.target)
  //入力データを取りたいけど.getElementById("input-data")で取れない
  //ここで取らなくてもいい？post_channel.rbのdata['text']['text']で取れる？
  obj.message = event.target.value //「e.target.value」でテキストボックスに打ち込んだ文字列を取得
  obj.user_id = event.target.getAttribute("data-user_id")
  // ↑ inputタグの属性を指定して値を取得している
  obj.post_id = event.target.getAttribute("data-post_id")
  obj.name = event.target.getAttribute("data-name")
  // console.log(obj.user_id)
  console.log(obj.message)
  if (event.keyCode === 13) {  //Enterキーが押下されたことを表す
    chatChannel.speak(obj);    //post_channel.jsのspeakアクションを発火   何が入る？ -> obj.message
    event.target.value = '';  

    //console.log(event.target)  //  -> <input type="text"...class data-behavior="post_speaker" data-user_id="1">が出力される
    return event.preventDefault();
  }
});

//ここでいろいろな属性を持ったコメントが来ている。その属性をobjに代入した。